provider/core/src/response.rs

```diff
@@ -74,38 +69,68 @@ pub struct DataResponseMetadata {
 pub struct DataPayload<M: DataMarker>(pub(crate) DataPayloadInner<M>);
 
 pub(crate) enum DataPayloadInner<M: DataMarker> {
-    Yoke(Yoke<M::Yokeable, Option<Cart>>),
+    Yoke(Yoke<M::Yokeable, Cart>),
     StaticRef(&'static M::Yokeable),
 }
 
 /// The type of the "cart" that is used by `DataPayload`.
 #[derive(Clone, Debug)]
 #[allow(clippy::redundant_allocation)] // false positive, it's cheaper to wrap an existing Box in an Rc than to reallocate a huge Rc
-pub struct Cart(SelectedRc<Box<[u8]>>);
+pub struct Cart(OptionSelectedRcBytes);
 
-impl Deref for Cart {
-    type Target = Box<[u8]>;
-    fn deref(&self) -> &Self::Target {
-        &self.0
-    }
-}
-// Safe because both Rc and Arc are StableDeref, and our impl delegates.
-unsafe impl stable_deref_trait::StableDeref for Cart {}
-// Safe because both Rc and Arc are CloneableCart, and our impl delegates.
+// Safe because OptionSelectedRcBytes implements CloneableCart
 unsafe impl yoke::CloneableCart for Cart {}
 
 impl Cart {
     /// Creates a `Yoke<Y, Option<Cart>>` from owned bytes by applying `f`.
+    #[deprecated]
     pub fn try_make_yoke<Y, F, E>(cart: Box<[u8]>, f: F) -> Result<Yoke<Y, Option<Self>>, E>
     where
         for<'a> Y: Yokeable<'a>,
         F: FnOnce(&[u8]) -> Result<<Y as Yokeable>::Output, E>,
     {
         Yoke::try_attach_to_cart(SelectedRc::new(cart), |b| f(b))
             // Safe because the cart is only wrapped
-            .map(|yoke| unsafe { yoke.replace_cart(Cart) })
+            .map(|yoke| unsafe {
+                yoke.replace_cart(|selected_rc| {
```

Comment:
nit: we should have a more detailed safety comment now that this is not a simple wrapping

Reply:
things that are important:

 - from_selected_rc does not panic
 - from_selected_rc does not ever destroy its parameter
 - from_selected_rc returns an object which owns its parameter

Reply:
also below